kernel/src/table_changes/scan_file.rs

```diff
@@ -0,0 +1,447 @@
+//! This module handles [`CDFScanFile`]s for [`TableChangeScan`]. A [`CDFScanFile`] consists of all the
+//! metadata required to generate a change data feed. [`CDFScanFile`] can be constructed using
+//! [`CDFScanFileVisitor`]. The visitor reads from engine data with the schema [`cdf_scan_row_schema`].
+//! You can convert engine data to this schema using the [`get_cdf_scan_row_expression`].
+use itertools::Itertools;
+use std::collections::HashMap;
+use std::sync::{Arc, LazyLock};
+
+use super::log_replay::TableChangesScanData;
+use crate::actions::visitors::visit_deletion_vector_at;
+use crate::engine_data::{GetData, TypedGetData};
+use crate::expressions::{column_expr, Expression};
+use crate::scan::state::DvInfo;
+use crate::schema::{
+    ColumnName, ColumnNamesAndTypes, DataType, MapType, SchemaRef, StructField, StructType,
+};
+use crate::utils::require;
+use crate::{DeltaResult, EngineData, Error, RowVisitor};
+
+/// Transforms an iterator of TableChangesScanData into an iterator  of
+/// `(CDFScanFile, Option<Arc<HashMap<String, DvInfo>>>)`. The `CDFScanFile` is constructed by
+/// visiting the engine data.
+#[allow(unused)]
+pub(crate) fn scan_data_to_scan_file(
+    scan_data: impl Iterator<Item = DeltaResult<TableChangesScanData>>,
+) -> impl Iterator<Item = DeltaResult<(CDFScanFile, Option<Arc<HashMap<String, DvInfo>>>)>> {
+    scan_data
+        .map(|scan_data| -> DeltaResult<_> {
+            let scan_data = scan_data?;
+            let callback: CDFScanCallback<Vec<CDFScanFile>> =
+                |context, scan_file| context.push(scan_file);
+            let result = visit_cdf_scan_files(
+                scan_data.data.as_ref(),
+                &scan_data.selection_vector,
+                vec![],
+                callback,
+            )?
+            .into_iter()
+            .map(move |scan_file| (scan_file, scan_data.remove_dv.clone()));
+            Ok(result)
+        }) // Iterator<DeltaResult<Iterator<(CDFScanFile, Option<Arc<HashMap<String, DvInfo>>>)>>>
+        .flatten_ok() // Iterator<DeltaResult<(CDFScanFile, Option<Arc<HashMap<String, DvInfo>>>)>>
```

Comment:
nit: IMO it's actually harder to read these comments because they are "too" detailed. All that's really needed is to give a hint of how we simplify results and iterators:
```suggestion
        }) // Iterator-Result-Iterator
        .flatten_ok() // Iterator-Result
```
(if we actually want to give the full type, it's probably better to do an actual type annotation so the compiler can verify it... which seems like overkill given that we're just trying to document the effect of e.g. `flatten_ok`)